File: backend/app/variants.py
"""Working out which form of a product is being bought.

A product is sold as one or more variants, and every line in a cart or an
order names one of them. Requests that predate variants do not, so a product
with a single variant resolves to it: that keeps older clients working without
a second code path, while a product with a real choice insists on being told.
"""

from typing import List, Optional

from .errors import FieldError


def variants_of(product: dict) -> List[dict]:
    return product.get("variants") or []


def find_variant(product: dict, variant_id: str) -> Optional[dict]:
    return next(
        (variant for variant in variants_of(product) if variant.get("variant_id") == variant_id),
        None,
    )


def resolve_variant(product: dict, variant_id: Optional[str]) -> dict:
    """The variant being bought, or a refusal explaining what is missing."""
    variants = variants_of(product)
    name = product.get("name", "This product")

    if not variants:
        # Every product is given a variant when it is created, so this means
        # a record written before that rule existed and never migrated.
        raise FieldError(409, "variant_id", f"{name} is not available to buy.")

    if variant_id:
        variant = find_variant(product, variant_id)
        if not variant:
            raise FieldError(404, "variant_id", f"That option of {name} is no longer available.")
        return variant

    if len(variants) == 1:
        return variants[0]

    option = product.get("option_name", "option").lower()
    raise FieldError(422, "variant_id", f"Choose a {option} for {name}.")


def line_key(product_id: str, variant_id: str) -> tuple:
    """What makes a cart or order line unique.

    Two sizes of the same product are two lines, not one, so the identity has
    to carry both parts.
    """
    return (product_id, variant_id)

File: frontend/src/lib/variants.js
/**
 * Which form of a product the shop is talking about.
 *
 * A product is sold as one or more variants: a weight for powders, a volume
 * for liquids, a colour for soaps. Price, stock and often the photograph
 * belong to the form rather than the product, so almost every screen needs to
 * know which one it is showing.
 */

export function variantsOf(product) {
  return product?.variants || [];
}

export function findVariant(product, variantId) {
  return variantsOf(product).find((variant) => variant.variant_id === variantId) || null;
}

/** Whether the customer has a decision to make. */
export function hasChoice(product) {
  return variantsOf(product).length > 1;
}

/**
 * The form to show before anyone has chosen.
 *
 * The cheapest in stock, because that is the price the grid advertises and
 * landing on a sold-out form asks the customer to fix something they did not
 * do. Everything sold out falls back to the cheapest of those, so the page
 * still has something to show.
 */
export function defaultVariant(product) {
  const byPrice = [...variantsOf(product)].sort((a, b) => (a.price || 0) - (b.price || 0));
  return byPrice.find((variant) => (variant.stock || 0) > 0) || byPrice[0] || null;
}

export function cheapestPrice(product) {
  const prices = variantsOf(product).map((variant) => variant.price);
  return prices.length ? Math.min(...prices) : undefined;
}

/**
 * The form a product is advertised at: the cheapest on offer.
 *
 * The grid shows this one's price and its own MRP. Pairing one form's price
 * with another's MRP would invent a discount neither of them offers.
 */
export function cheapestVariant(product) {
  return variantsOf(product).reduce(
    (best, variant) => (!best || (variant.price || 0) < (best.price || 0) ? variant : best),
    null
  );
}

/** Everything on hand across a product's forms. */
export function totalStock(product) {
  const variants = variantsOf(product);
  if (!variants.length) return undefined;
  return variants.reduce((total, variant) => total + (variant.stock || 0), 0);
}

/**
 * A product as it is actually sold in one form.
 *
 * The price, stock and image of the chosen form stand in for the product's
 * own, under the names every screen already reads, so the pages below do not
 * each have to remember which is which.
 */
export function asSold(product, variant) {
  if (!product || !variant) return product;
  return {
    ...product,
    price: variant.price,
    mrp: variant.mrp ?? variant.price,
    stock: variant.stock ?? 0,
    image: variant.image || product.image,
    variant_id: variant.variant_id,
    variant_label: variant.label,
  };
}

/**
 * What makes a cart line unique.
 *
 * Two forms of one product are two lines, so the identity has to carry both
 * parts. It doubles as a React key and a test handle.
 */
export function lineKey(item) {
  if (!item) return "";
  return item.variant_id ? `${item.product_id}--${item.variant_id}` : item.product_id;
}

/** How a line is named once the product alone is not enough. */
export function lineName(item) {
  return item?.variant_label ? `${item.name} (${item.variant_label})` : item?.name;
}
